File: Tecniques de Programació Avançades/Practica 1/src/WordCount.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;
import java.util.function.Function;

public class WordCount {

    public static void main(String[] args) {
        Controller<String, Map<String, Integer>> mapController = new Controller<>(4, 1024, 10, new RoundRobin());
        Controller<Map<String, Integer>, Integer> reduceController = new Controller<>(4, 1024, 10, new RoundRobin());

        List<String> textFiles = obtenerTextosDesdeGutenberg();

        // Fase de map: invocación grupal
        Function<String, Map<String, Integer>> funcionMap = WordCount::map;
        mapController.registerAction("map", funcionMap, 256);
        List<Map<String, Integer>> resultados = mapController.invoke("map", textFiles, 256);

        // Fase de reduce: invocación individual
        Function<Map<String, Integer>, Integer> funcionReduce = wordCountMap -> {
            Map<String, Integer> contar = new HashMap<>();

            wordCountMap.forEach((word, count) -> {
                String palabralimpia = word.replaceAll("[^a-zA-Z]", "").toLowerCase();

                if (contar.containsKey(palabralimpia)) {
                    contar.put(palabralimpia, contar.get(palabralimpia) + count);
                } else {
                    contar.put(palabralimpia, count);
                }
            });

            contar.forEach((word, count) -> {
                System.out.println("Veces repetida " + word + ": " + count);
            });

            return wordCountMap.values().stream().mapToInt(Integer::intValue).sum();
        };

        reduceController.registerAction("reduce", funcionReduce, 256);

        // Invocar la fase de reduce con los resultados de la fase de map
        for (int i = 0; i < resultados.size(); i++) {
            Integer result = reduceController.invokeS("reduce", resultados.get(i), 256);
            System.out.println("Resultado para el archivo " + i + ": " + result);
        }
    }

    public static Map<String, Integer> map(String input) {
        Map<String, Integer> wordCounts = new HashMap<>();
        StringTokenizer tokenizer = new StringTokenizer(input, " \t\n\r\f,.:;?![]'\"()");

        while (tokenizer.hasMoreTokens()) {
            String word = tokenizer.nextToken().toLowerCase();
            if (wordCounts.containsKey(word)) {
                wordCounts.put(word, wordCounts.get(word) + 1);
            } else {
                wordCounts.put(word, 1);
            }
        }

        return wordCounts;
    }

    private static List<String> obtenerTextosDesdeGutenberg() {
        try {
            Path directory = Paths.get(System.getProperty("user.dir"));
            return Files.walk(directory)
                        .filter(Files::isRegularFile)
                        .filter(file -> file.toString().endsWith(".txt"))
                        .map(file -> {
                            try {
                                return new String(Files.readAllBytes(file));
                            } catch (Exception e) {
                                e.printStackTrace();
                                return "";
                            }
                        })
                        .toList();
        } catch (Exception e) {
            e.printStackTrace();
            return List.of();
        }
    }
}

File: Tecniques de Programació Avançades/Practica 2 - Aleix Carrillo i Oriol Poblet/src/main/java/WordCount.scala
import java.nio.file.{Files, Paths}
import java.util.StringTokenizer
import scala.collection.mutable.HashMap
import scala.jdk.CollectionConverters._

object WordCount {
    def main(args: Array[String]): Unit = {
        val controller = new Controller[Map[String, Int], Int](4, 1024, 10, new RoundRobin)

        val textFiles = obtenerTextosDesdeGutenberg()
        val input = textFiles.map(CountWords.map)

        val mapAction: Function[Map[String, Int], Int] = (wordCountMap: Map[String, Int]) => {
            wordCountMap.foreach { case (word, count) =>
                println(s"Veces repetida $word: $count")
            }
            wordCountMap.values.sum
        }

        controller.registerAction("mapAction", mapAction, 256)

        val result = controller.invoke("mapAction", input, 256)

        val Titulos = obtenerNombresArchivosDesdeGutenberg()

        (Titulos, result).zipped.foreach { (title, count) =>
            println(s"Palabras para $title: $count")
        }
    }

    def map(input: String): Map[String, Int] = {
        val wordCounts = new HashMap[String, Int]()
        val tokenizer = new StringTokenizer(input)

        while (tokenizer.hasMoreTokens) {
            val word = tokenizer.nextToken().toLowerCase
            wordCounts.update(word, wordCounts.getOrElse(word, 0) + 1)
        }
        wordCounts.toMap
    }

    def obtenerTextosDesdeGutenberg(): List[String] = {
        val directory = Paths.get(System.getProperty("user.dir"))
        Files.walk(directory)
                .iterator()
                .asScala
                .filter(Files.isRegularFile(_))
                .filter(_.toString.endsWith(".txt"))
                .map(file => new String(Files.readAllBytes(file)))
                .toList
    }

    def obtenerNombresArchivosDesdeGutenberg(): List[String] = {
        val directory = Paths.get(System.getProperty("user.dir"))
        Files.walk(directory)
                .iterator()
                .asScala
                .filter(Files.isRegularFile(_))
                .filter(_.toString.endsWith(".txt"))
                .map(_.getFileName.toString)
                .toList
    }
}